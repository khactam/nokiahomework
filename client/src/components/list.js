import React, { Component, useState } from 'react';
import { AgGridColumn, AgGridReact } from 'ag-grid-react';
import 'ag-grid-community/dist/styles/ag-grid.css';
import 'ag-grid-community/dist/styles/ag-theme-alpine.css';

class ListData extends Component {
    constructor(props) {
        super(props);
        this.state = {
            rowData: []
        };
    }
    componentDidMount() {
        this.callBackendAPI()
            .then(res => this.setState({ rowData: res.response }))
            .catch(err => console.log(err));
    }
    // fetching the GET route from the Express server which matches the GET route from server.js
    callBackendAPI = async () => {
        const response = await fetch('/listdata');
        const body = await response.json();
        if (response.status !== 200) {
            throw Error(body.message)
        }
        return body;
    };

    cellRenderer = (params) => {
        let image = null;
        let textValue = null;
        switch (params.data.status) {
            case 1:
                image = 'finished.svg';
                textValue = 'Finished';
                break;
            case 2:
                image = 'failed.svg';
                textValue = 'Failed';
                break;
            case 3:
                image = 'interrupted.svg';
                textValue = 'Interrupted';
                break;
            default:
                break;
        }
        const imageSource = `./assets/${image}`;
        if(textValue && image) {
            return (
                `<span>
                    <img src=${imageSource} style="width:20px"/>
                    ${textValue}
                </span>`
            )
        }
        else return ''
    }

    render() {
        return <div className="ag-theme-alpine" style={{ height: '100vh', width: '100vw' }} defaultcoldef={{
            flex: 1,
        }}>
            <AgGridReact
                rowData={this.state.rowData} defaultColDef={{
                    flex: 1,
                    minWidth: 150,
                    filter: true,
                    sortable: true
                }}
                suppressRowTransform={true}>
                <AgGridColumn field="operation"></AgGridColumn>
                <AgGridColumn field="scope"></AgGridColumn>
                <AgGridColumn field="timeStamp"></AgGridColumn>
                <AgGridColumn field="status" cellRenderer={this.cellRenderer}></AgGridColumn>
            </AgGridReact>
        </div>
    };
}

export default ListData;