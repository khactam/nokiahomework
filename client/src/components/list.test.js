import { render, screen } from '@testing-library/react';
import { ReactDOM } from 'react';
import renderer from 'react-test-renderer';
import { cellRenderer } from './list'
import { shallow, mount } from 'enzyme'
import ListData from './list'

describe('ListData should display correctly', () => {
    let wrapper, listData, agGridReact;
    beforeAll(() => {
        wrapper = shallow(<ListData />);
        listData = mount(<ListData />);
        agGridReact = listData.find('AgGridReact').instance();
    });


    it('renders ListData without crashing', () => {
        expect(wrapper.exists()).toBe(true);
    });
    it('has table with operation column', async () => {
        expect(listData.find('div').first().text()).toContain('Operation');
    });
    it('checking function componentDidMount is called and callBackendAPI ok', () => {
        const instance = listData.instance();
        const spy = jest.spyOn(instance, 'callBackendAPI');
        instance.componentDidMount();
        expect(spy).toHaveBeenCalled();
    })
    it('renders some operations', async () => {

    })
    test('statusCellRenderer function ok', () => {
        wrapper.instance().statusCellRenderer = jest.fn();
        wrapper.update();
        expect(wrapper.instance().statusCellRenderer).toMatchSnapshot();
    });
    test('statusCellRenderer renders correct status Finished in cell', () => {
        const fakeData = {
            data: {
                status: 1
            }
        }
        const testwrapper = shallow(<ListData />)
        const test = testwrapper.instance().cellRenderer(fakeData);
        expect(test.includes('Finished')).toBe(true);
        expect(test.includes('./assets/finished.svg')).toBe(true);
    })
    test('statusCellRenderer renders correct status Failed in cell', () => {
        const fakeData = {
            data: {
                status: 2
            }
        }
        const testwrapper = shallow(<ListData />)
        const test = testwrapper.instance().cellRenderer(fakeData);
        expect(test.includes('Failed')).toBe(true);
        expect(test.includes('./assets/failed.svg')).toBe(true);
    })
    test('statusCellRenderer renders correct status Interrupted in cell', () => {
        const fakeData = {
            data: {
                status: 3
            }
        }
        const testwrapper = shallow(<ListData />)
        const test = testwrapper.instance().cellRenderer(fakeData);
        expect(test.includes('Interrupted')).toBe(true);
        expect(test.includes('./assets/interrupted.svg')).toBe(true);
    })
    test('statusCellRenderer renders nothing', () => {
        const fakeData = {
            data: {
            }
        }
        const testwrapper = shallow(<ListData />)
        const test = testwrapper.instance().cellRenderer(fakeData);
        expect(test).toEqual('');
    })
})