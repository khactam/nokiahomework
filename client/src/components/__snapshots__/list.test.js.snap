// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListData should display correctly statusCellRenderer function ok 1`] = `[MockFunction]`;
